linesearch.go

```diff
@@ -10,132 +10,135 @@ import (
 	"github.com/gonum/floats"
 )
 
-// Linesearch is a linesearch-based optimization method.
-// It consists of a NextDirectioner, which specifies the next linesearch method,
-// and a LinesearchMethod which performs the linesearch in the direction specified
-// by the NextDirectioner.
-type Linesearch struct {
+// LinesearchMethod encapsulates the common functionality of gradient-based
+// line-search methods and can serve as a helper struct for their
+// implementation. It does not implement the Method interface because of the
+// missing Needs() method. It consists of a NextDirectioner, which specifies
+// the search direction at each iteration, and a Linesearcher which performs a
+// linesearch along the search direction.
+type LinesearchMethod struct {
 	NextDirectioner NextDirectioner
-	Method          LinesearchMethod
+	Linesearcher    Linesearcher
 
-	initX []float64
-	dir   []float64
+	x     []float64 // Starting point for the current iteration.
+	dir   []float64 // Search direction for the current iteration.
+	first bool      // Indicator of the first iteration.
 
-	lastEvalType EvaluationType
-	iterType     IterationType
+	evalType EvaluationType
+	iterType IterationType
 }
 
-func (ls *Linesearch) Init(loc *Location, xNext []float64) (EvaluationType, IterationType, error) {
-	ls.initX = resize(ls.initX, len(loc.X))
-	copy(ls.initX, loc.X)
+func (ls *LinesearchMethod) Init(loc *Location, xNext []float64) (EvaluationType, IterationType, error) {
+	if loc.Gradient == nil {
+		panic("linesearch: gradient is nil")
+	}
 
-	ls.dir = resize(ls.dir, len(loc.X))
-	stepSize := ls.NextDirectioner.InitDirection(loc, ls.dir)
+	dim := len(loc.X)
+	ls.x = resize(ls.x, dim)
+	ls.dir = resize(ls.dir, dim)
+	ls.first = true
 
-	projGrad := math.NaN()
-	if loc.Gradient != nil {
-		projGrad = floats.Dot(loc.Gradient, ls.dir)
-		if projGrad >= 0 {
-			return NoEvaluation, NoIteration, ErrNonNegativeStepDirection
-		}
-	}
-	lsLoc := LinesearchLocation{
-		F:          loc.F,
-		Derivative: projGrad,
-	}
-	evalType := ls.Method.Init(lsLoc, stepSize)
-	floats.AddScaledTo(xNext, ls.initX, stepSize, ls.dir)
-	ls.lastEvalType = evalType
-	ls.iterType = MinorIteration
-	return evalType, ls.iterType, nil
+	return ls.initNextLinesearch(loc, xNext)
 }
 
-func (ls *Linesearch) Iterate(loc *Location, xNext []float64) (EvaluationType, IterationType, error) {
+func (ls *LinesearchMethod) Iterate(loc *Location, xNext []float64) (EvaluationType, IterationType, error) {
+	if ls.iterType == NoIteration {
```

Comment:
Without calling Init() the only thing the client can do to recover from no progress is something with the Linesearcher. With this check it is not possible, so I will remove it, there is not much harm in doing that.